File: go_app/internal/ui/rapid_diffs/diff_file_header.go
package rapid_diffs

import (
	"encoding/json"
	"fmt"
	"strings"
)

// DiffFileHeaderComponent represents a diff file header component
type DiffFileHeaderComponent struct {
	DiffFile *DiffFile
}

// NewDiffFileHeaderComponent creates a new diff file header component
func NewDiffFileHeaderComponent(diffFile *DiffFile) *DiffFileHeaderComponent {
	return &DiffFileHeaderComponent{
		DiffFile: diffFile,
	}
}

// GetOptionsMenuItems returns the options menu items
func (d *DiffFileHeaderComponent) GetOptionsMenuItems() string {
	viewTitle := fmt.Sprintf("View file @ %s", d.DiffFile.ContentSha[:8])
	viewHref := fmt.Sprintf("/projects/%d/blob/%s/%s",
		d.DiffFile.Repository.Project.ID,
		d.DiffFile.ContentSha,
		d.DiffFile.NewPath)

	items := []map[string]string{
		{
			"text": viewTitle,
			"href": viewHref,
		},
	}

	jsonBytes, _ := json.Marshal(items)
	return string(jsonBytes)
}

// RenderCopyPathButton renders the copy path button
func (d *DiffFileHeaderComponent) RenderCopyPathButton() string {
	return fmt.Sprintf(`
		<button class="gl-button btn btn-default btn-icon btn-sm"
			data-clipboard-text="%s"
			data-clipboard-gfm="\`%s\`"
			title="Copy file path"
			data-placement="top"
			data-boundary="viewport"
			data-testid="rd-diff-file-copy-clipboard">
			<span class="gl-icon">copy</span>
		</button>
	`, d.DiffFile.FilePath, d.DiffFile.FilePath)
}

// Render generates the HTML for the diff file header component
func (d *DiffFileHeaderComponent) Render() string {
	var parts []string

	parts = append(parts, `<div class="rd-diff-file-header" data-testid="rd-diff-file-header">`)

	// Toggle buttons
	parts = append(parts, `<div class="rd-diff-file-toggle">`)
	parts = append(parts, `
		<button class="gl-button btn btn-default btn-sm rd-diff-file-toggle-button" data-opened data-click="toggleFile" aria-label="Hide file contents">
			<span class="gl-icon">chevron-down</span>
		</button>
		<button class="gl-button btn btn-default btn-sm rd-diff-file-toggle-button" data-closed data-click="toggleFile" aria-label="Show file contents">
			<span class="gl-icon">chevron-right</span>
		</button>
	`)
	parts = append(parts, "</div>")

	// Title section
	parts = append(parts, `<div class="rd-diff-file-title">`)
	if d.DiffFile.Submodule {
		parts = append(parts, fmt.Sprintf(`
			<span data-testid="rd-diff-file-header-submodule">
				<span class="gl-icon gl-file-icon">folder-git</span>
				<strong>%s</strong>
			</span>
			%s
		`, d.DiffFile.FilePath, d.RenderCopyPathButton()))
	} else {
		if d.DiffFile.RenamedFile {
			parts = append(parts, fmt.Sprintf(`
				<strong>%s</strong>
				→
				<strong>%s</strong>
			`, d.DiffFile.OldPath, d.DiffFile.NewPath))
		} else {
			parts = append(parts, fmt.Sprintf(`<strong>%s</strong>`, d.DiffFile.FilePath))
			if d.DiffFile.DeletedFile {
				parts = append(parts, " deleted")
			}
			parts = append(parts, d.RenderCopyPathButton())
		}

		if d.DiffFile.ModeChanged {
			parts = append(parts, fmt.Sprintf(`<small>%s → %s</small>`, d.DiffFile.AMode, d.DiffFile.BMode))
		}

		if d.DiffFile.StoredExternally && d.DiffFile.ExternalStorage == "lfs" {
			parts = append(parts, `<span class="gl-badge badge badge-neutral">LFS</span>`)
		}
	}
	parts = append(parts, "</div>")

	// Info section
	parts = append(parts, `<div class="rd-diff-file-info">`)
	parts = append(parts, `<div class="rd-diff-file-stats">`)
	parts = append(parts, fmt.Sprintf(`
		<span class="rd-lines-added">
			<span>+</span>
			<span data-testid="js-file-addition-line">%d</span>
		</span>
		<span class="rd-lines-removed">
			<span>−</span>
			<span data-testid="js-file-deletion-line">%d</span>
		</span>
	`, d.DiffFile.AddedLines, d.DiffFile.RemovedLines))
	parts = append(parts, "</div>")

	// Options menu
	parts = append(parts, `<div class="rd-diff-file-options-menu">`)
	parts = append(parts, `<div class="js-options-menu">`)
	parts = append(parts, fmt.Sprintf(`<script type="application/json">%s</script>`, d.GetOptionsMenuItems()))
	parts = append(parts, `
		<button class="gl-button btn btn-default btn-sm js-options-button" data-click="toggleOptionsMenu" aria-label="Options">
			<span class="gl-icon">ellipsis_v</span>
		</button>
	`)
	parts = append(parts, "</div>")
	parts = append(parts, "</div>")
	parts = append(parts, "</div>")

	parts = append(parts, "</div>")

	return strings.Join(parts, "\n")
}

File: rust_app/src/components/rapid_diffs/diff_file_header.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Represents a diff file header with additional metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiffFileHeaderData {
    /// The file path
    pub file_path: String,
    /// The old file path if renamed
    pub old_path: Option<String>,
    /// The new file path if renamed
    pub new_path: Option<String>,
    /// The content SHA
    pub content_sha: String,
    /// Whether the file is a submodule
    pub submodule: bool,
    /// Whether the file is renamed
    pub renamed_file: bool,
    /// Whether the file is deleted
    pub deleted_file: bool,
    /// Whether the file mode is changed
    pub mode_changed: bool,
    /// The old mode if changed
    pub a_mode: Option<String>,
    /// The new mode if changed
    pub b_mode: Option<String>,
    /// Whether the file is stored externally
    pub stored_externally: bool,
    /// The external storage type
    pub external_storage: Option<String>,
    /// The number of added lines
    pub added_lines: i32,
    /// The number of removed lines
    pub removed_lines: i32,
    /// The project ID
    pub project_id: i64,
}

/// Renders a diff file header component
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiffFileHeaderComponent {
    /// The diff file header data
    diff_file: DiffFileHeaderData,
}

impl DiffFileHeaderComponent {
    /// Create a new diff file header component
    pub fn new(diff_file: DiffFileHeaderData) -> Self {
        Self { diff_file }
    }

    /// Get the copy path button HTML
    fn copy_path_button(&self) -> String {
        format!(
            r#"<button class="btn btn-tertiary btn-sm" data-clipboard-text="{}" data-clipboard-gfm="`{}`" title="Copy file path" data-placement="top" data-boundary="viewport" data-testid="rd-diff-file-copy-clipboard">
                <i class="fa fa-copy"></i>
            </button>"#,
            self.diff_file.file_path, self.diff_file.file_path
        )
    }

    /// Get the options menu items JSON
    fn options_menu_items(&self) -> String {
        let view_title = format!("View file @ {}", self.diff_file.content_sha);
        let view_href = format!(
            "/projects/{}/repository/blobs/{}",
            self.diff_file.project_id, self.diff_file.content_sha
        );

        format!(r#"[{{"text":"{}","href":"{}"}}]"#, view_title, view_href)
    }

    /// Render the diff file header component
    pub fn render(&self) -> String {
        let mut html = String::new();

        // Add diff file header container
        html.push_str(r#"<div class="rd-diff-file-header" data-testid="rd-diff-file-header">"#);

        // Add diff file toggle
        html.push_str(r#"<div class="rd-diff-file-toggle">"#);
        html.push_str(r#"<button class="btn btn-tertiary btn-sm rd-diff-file-toggle-button" data-opened="" data-click="toggleFile" aria-label="Hide file contents"><i class="fa fa-chevron-down"></i></button>"#);
        html.push_str(r#"<button class="btn btn-tertiary btn-sm rd-diff-file-toggle-button" data-closed="" data-click="toggleFile" aria-label="Show file contents"><i class="fa fa-chevron-right"></i></button>"#);
        html.push_str("</div>");

        // Add diff file title
        html.push_str(r#"<div class="rd-diff-file-title">"#);

        if self.diff_file.submodule {
            // Submodule
            html.push_str(r#"<span data-testid="rd-diff-file-header-submodule">"#);
            html.push_str(r#"<i class="fa fa-folder"></i>"#);
            html.push_str(&format!(r#"<strong>{}</strong>"#, self.diff_file.file_path));
            html.push_str("</span>");
            html.push_str(&self.copy_path_button());
        } else {
            // Regular file
            if self.diff_file.renamed_file {
                // Renamed file
                if let (Some(old_path), Some(new_path)) =
                    (&self.diff_file.old_path, &self.diff_file.new_path)
                {
                    html.push_str(&format!(r#"<strong>{}</strong>"#, old_path));
                    html.push_str(" → ");
                    html.push_str(&format!(r#"<strong>{}</strong>"#, new_path));
                }
            } else {
                // Regular file
                html.push_str(&format!(r#"<strong>{}</strong>"#, self.diff_file.file_path));
                if self.diff_file.deleted_file {
                    html.push_str(" deleted");
                }
            }
            html.push_str(&self.copy_path_button());

            // Mode changed
            if self.diff_file.mode_changed {
                if let (Some(a_mode), Some(b_mode)) =
                    (&self.diff_file.a_mode, &self.diff_file.b_mode)
                {
                    html.push_str(&format!(r#"<small>{} → {}</small>"#, a_mode, b_mode));
                }
            }

            // External storage
            if self.diff_file.stored_externally {
                if let Some(storage) = &self.diff_file.external_storage {
                    if storage == "lfs" {
                        html.push_str(r#"<span class="badge badge-neutral">LFS</span>"#);
                    }
                }
            }
        }

        html.push_str("</div>");

        // Add diff file info
        html.push_str(r#"<div class="rd-diff-file-info">"#);

        // Add diff file stats
        html.push_str(r#"<div class="rd-diff-file-stats">"#);
        html.push_str(r#"<span class="rd-lines-added"><span>+</span><span data-testid="js-file-addition-line">"#);
        html.push_str(&format!("{}</span></span>", self.diff_file.added_lines));
        html.push_str(r#"<span class="rd-lines-removed"><span>−</span><span data-testid="js-file-deletion-line">"#);
        html.push_str(&format!("{}</span></span>", self.diff_file.removed_lines));
        html.push_str("</div>");

        // Add diff file options menu
        html.push_str(r#"<div class="rd-diff-file-options-menu">"#);
        html.push_str(r#"<div class="js-options-menu">"#);
        html.push_str(&format!(
            r#"<script type="application/json">{}</script>"#,
            self.options_menu_items()
        ));
        html.push_str(r#"<button class="btn btn-tertiary btn-sm js-options-button" data-click="toggleOptionsMenu" aria-label="Options"><i class="fa fa-ellipsis-v"></i></button>"#);
        html.push_str("</div></div>");

        // Close diff file info
        html.push_str("</div>");

        // Close diff file header
        html.push_str("</div>");

        html
    }
}
